modules/GenericPowermeter/main/powermeterImpl.cpp

```diff
@@ -325,216 +324,245 @@ powermeterImpl::ModbusFunctionType powermeterImpl::select_modbus_function(const
+        this->pm_last_values.current_A = amp;
+    } else if (register_data.type == CURRENT_A_L2) {
+        types::units::Current amp = this->pm_last_values.current_A.value();
+        amp.L2 = this->merge_register_values_into_element(register_data, exponent, register_message);
+        this->pm_last_values.current_A = amp;
+    } else if (register_data.type == CURRENT_A_L3) {
+        types::units::Current amp = this->pm_last_values.current_A.value();
+        amp.L3 = this->merge_register_values_into_element(register_data, exponent, register_message);
+        this->pm_last_values.current_A = amp;
+    } else if (register_data.type == FREQUENCY_HZ_L1) {
+        types::units::Frequency freq = this->pm_last_values.frequency_Hz.value();
+        freq.L1 = this->merge_register_values_into_element(register_data, exponent, register_message);
+        this->pm_last_values.frequency_Hz = freq;
+    } else if (register_data.type == FREQUENCY_HZ_L2) {
+        types::units::Frequency freq = this->pm_last_values.frequency_Hz.value();
+        freq.L2 = this->merge_register_values_into_element(register_data, exponent, register_message);
+        this->pm_last_values.frequency_Hz = freq;
+    } else if (register_data.type == FREQUENCY_HZ_L3) {
+        types::units::Frequency freq = this->pm_last_values.frequency_Hz.value();
+        freq.L3 = this->merge_register_values_into_element(register_data, exponent, register_message);
+        this->pm_last_values.frequency_Hz = freq;
+    } else {
+    }
+
+    return true;
 }
 
 float powermeterImpl::merge_register_values_into_element(const RegisterData& reg_data, const int16_t exponent,
                                                          const types::serial_comm_hub_requests::Result& reg_message) {
-    if (reg_message.value.has_value()) {
-        uint32_t value{0};
-        auto& reg_value = reg_message.value.value();
-        if (reg_data.num_registers == 1 && reg_value.size() == 1) {
-            value = reg_value.at(0);
-        } else if (reg_data.num_registers == 2 && reg_value.size() == 2) {
-            value += reg_value.at(0) << 16;
-            value += reg_value.at(1);
-        } else {
-            throw std::runtime_error("Values of more than 2 registers in size are currently not supported!");
-        }
-
-        auto val = *reinterpret_cast<float*>(&value);
-        auto val_scaled = float(val * reg_data.multiplier * pow(10.0, exponent));
-
-        return val_scaled;
+    uint32_t value{0};
+    const auto& reg_value = reg_message.value.value();
+    if (reg_data.num_registers == 1 && reg_value.size() == 1) {
+        value = reg_value.at(0);
+    } else if (reg_data.num_registers == 2 && reg_value.size() == 2) {
+        value += reg_value.at(0) << 16;
+        value += reg_value.at(1);
     } else {
-        EVLOG_error << "Error! Received message is empty!\n";
+        throw std::runtime_error("Values of more than 2 registers in size are currently not supported!");
```

Comment:
In general it would be great to move towards proper error handling with the error framework (like in other powermeters in the codebase). But since this isn't a regression this doesn't have to be addressed in this PR